### PR TITLE
[MDEP-644] Reactor functionality added for verbose tree

### PR DESCRIPTION
**Pull request by ian-lavallee**
Adds multi module support

### CODE REVIEW

**Comment by ian-lavallee** [comment]
@elharo Reactor PR recreated. It was easier to just make a new branch with the changes than rebase.

File: src/main/java/org/apache/maven/plugins/dependency/tree/VerboseGraphTgfSerializer.java [hunk 1]
```diff
@@ -49,7 +49,16 @@ public String serialize( DependencyNode root )
         Artifact rootArtifact = root.getArtifact();
         nodes.append( root.hashCode() ).append( " " ).append( rootArtifact.getGroupId() ).append( ":" ).append(
                 rootArtifact.getArtifactId() ).append( ":" ).append( rootArtifact.getExtension() ).append( ":" ).append(
-                rootArtifact.getVersion() ).append( "\n" );
+                rootArtifact.getVersion() );
+
+        if ( root.getData().containsKey( "ContainsModule" ) )
+        {
+            nodes.append( " WARNING: this tree contains a submodule and once it reaches the submodule will print "
```
Comment (bimargulies-google) [inline src/main/java/org/apache/maven/plugins/dependency/tree/VerboseGraphTgfSerializer.java:56]: this is a little hard to follow. a semicolon after the first 'submodule' would help.

File: src/main/java/org/apache/maven/plugins/dependency/tree/VerboseGraphTextSerializer.java [hunk 2]
```diff
@@ -45,8 +45,16 @@ public String serialize( DependencyNode root )
         // deal with root first
         Artifact rootArtifact = root.getArtifact();
         builder.append( rootArtifact.getGroupId() ).append( ":" ).append( rootArtifact.getArtifactId() ).append(
-                ":" ).append( rootArtifact.getExtension() ).append( ":" ).append( rootArtifact.getVersion() ).append(
-                "\n" );
+                ":" ).append( rootArtifact.getExtension() ).append( ":" ).append( rootArtifact.getVersion() );
+
+        if ( root.getData().containsKey( "ContainsModule" ) )
+        {
+            builder.append( " WARNING: this tree contains a submodule and once it reaches the submodule will print "
```
Comment (bimargulies-google) [inline src/main/java/org/apache/maven/plugins/dependency/tree/VerboseGraphTextSerializer.java:52]: I'd write:

contains a submodule; once it reaches

File: src/main/java/org/apache/maven/plugins/dependency/tree/VerboseGraphDotSerializer.java [hunk 3]
```diff
@@ -49,7 +49,16 @@ public String serialize( DependencyNode root )
         StringBuilder result = new StringBuilder( "digraph" );
         result.append( " \"" ).append( rootArtifact.getGroupId() ).append( ":" ).append(
                 rootArtifact.getArtifactId() ).append( ":" ).append( rootArtifact.getExtension() ).append( ":" ).append(
-                rootArtifact.getVersion() ).append( "\" {" ).append( "\n" );
+                rootArtifact.getVersion() );
+
+        if ( root.getData().containsKey( "ContainsModule" ) )
+        {
+            result.append( " WARNING: this tree contains a submodule and once it reaches the submodule will print "
```
Comment (bimargulies-google) [inline src/main/java/org/apache/maven/plugins/dependency/tree/VerboseGraphDotSerializer.java:56]: contains a submodule; once

File: src/it/projects/tree-verbose-multimodule/verify.bsh [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    throw new Exception( "Unexpected root dependency tree" );
+}
+
+actual = FileUtils.fileRead( new File( basedir, "module-a/target/tree.txt" ) );
+expected = FileUtils.fileRead( new File( basedir, "module-a/expected.txt" ) );
+
+actual = actual.replaceAll( "[\n\r]+", "\n" );
+expected = expected.replaceAll( "[\n\r]+", "\n" );
+
+System.out.println( "Checking module-a dependency tree..." );
```
Comment (elharo) [inline src/it/projects/tree-verbose-multimodule/verify.bsh:43]: System.out.println isn't usually desired in tests. Passing tests should generate no output.

File: src/main/java/org/apache/maven/plugins/dependency/tree/ArtifactKey.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+ */
+
+import org.apache.maven.project.MavenProject;
+
+/**
+ * Uniquely defines an artifact by groupId, artifactId and version.
+ */
+final class ArtifactKey
+{
+    private static final int HASH_PRIME = 31;
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/tree/ArtifactKey.java:29]: this can be a local variable

File: src/main/java/org/apache/maven/plugins/dependency/tree/VerboseGraphGraphmlSerializer.java [hunk 6]
```diff
@@ -152,6 +152,12 @@ private String getGraphmlNodeLine( DependencyNode node, Map<DependencyNode, Stri
                     "scope managed from " + node.getArtifact().getProperties().get( PRE_MANAGED_SCOPE );
         }
         builder.append( getDependencyCoordinate( node ) ).append( coordString );
+        if ( node.getData().containsKey( "ContainsModule" ) )
+        {
+            builder.append( " WARNING: this tree contains a submodule and once it reaches the submodule will print "
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/tree/VerboseGraphGraphmlSerializer.java:157]: submodule. Once

File: src/main/java/org/apache/maven/plugins/dependency/tree/VerboseGraphTextSerializer.java (hunk 2 above)
Comment (elharo) [reply src/main/java/org/apache/maven/plugins/dependency/tree/VerboseGraphTextSerializer.java:52]: or just convert to two sentences

**Comment by ian-lavallee** [comment]
@elharo can you merge this now please

**Comment by jira-importer** [comment]
Resolve #1078
